perspective.el

```diff
@@ -968,6 +990,249 @@ perspective beginning with the given letter."
+                                                             if (persp--state-interesting-buffer-p buffer)
+                                                             collect (buffer-name buffer)))
+                                                   (windows
+                                                    (cl-loop for entry in (window-state-get (frame-root-window) t)
+                                                             collect (persp--state-window-state-massage entry persp buffers))))
+                                      (puthash persp
+                                               (make-persp--state-single
+                                                :buffers buffers
+                                                :windows windows)
+                                               persps-in-frame)))))
+                       (make-persp--state-frame
+                        :persps persps-in-frame
+                        :order persp-names-in-order)))))
+
+;;;###autoload
+(cl-defun persp-state-save (&optional file interactive?)
+  "Save the current perspective state to FILE.
+
+FILE defaults to the value of persp-state-default-file if it is
+set.
+
+Each perspective's buffer list and window layout will be saved.
+Frames and their associated perspectives will also be saved
+(but not the original frame sizes).
+
+Buffers with * characters in their names, as well as buffers without
+associated files will be ignored. If such buffers are currently
+visible in a perspective as windows, they will be saved as
+'*scratch* (persp)' buffers."
+  (interactive (list
+                (read-file-name "Save perspective state to file: "
+                                persp-state-default-file
+                                persp-state-default-file)
+                t))
+  (unless persp-mode
+    (message "persp-mode not enabled, nothing to save")
+    (return-from persp-state-save))
+  (lexical-let ((target-file (if (and file (not (string-equal "" file)))
+                                 ;; file provided as argument, just use it
+                                 (expand-file-name file)
+                               ;; no file provided as argument
+                               (if interactive?
+                                   ;; return nil in interactive call mode, since
+                                   ;; read-file-name should have provided a reasonable
+                                   ;; default
+                                   nil
+                                 ;; in non-interactive call mode, we want to fall back to
+                                 ;; the default, but only if it is set
+                                 (if (and persp-state-default-file
+                                          (not (string-equal "" persp-state-default-file)))
+                                     (expand-file-name persp-state-default-file)
```

Comment:
I think both these `if`s can reduce to a single 

```
(unless (and (not interactive?)
             persp-state-default-file
             (not blah...blah))
  ...)
```

Reply:
hrm... that logic is probably a tad wrong... but `(if cond1 nil (if cond2 result nil))` can definitely be reduced to `(when-or-unless cond3 result)`

Reply:
Right. I went through several iterations of writing this logic, and had to keep it as explicit as possible to get it right. There are too many combinations for how the function could be called: interactively or not, with a file argument or not, and what is an omitted file argument in interactive mode with a default available anyway?

I'm not happy with the look of the Boolean algebra here, but I've also been called out for writing overly-clever conditions in the past. 🤷🏻‍♂️